### PR TITLE
Adding new app setting for key vault so we can set Vault URI

### PR DESCRIPTION
**Pull request by mohit-naroliya**
We're currently generating the Vault URI in the codebase using the keyvaultname, which works for Azure Cloud. However, since Azure Gov and Azure China Cloud use different Vault URIs, it doesn't work for them. I'm adding a new appsetting configuration that will allow us to specify the Vault URI directly.

### CODE REVIEW

**Comment by mohit-naroliya** [comment]
@dotnet-policy-service agree company="GovBuilt"

________________________________

File: src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs [hunk 1]
```diff
@@ -74,9 +74,15 @@ private static void AddOrchardCoreAzureKeyVault(
             throw new Exception("The 'KeyVaultName' property is no configured. Please configure it by specifying the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
         }
 
-        if (!Uri.TryCreate($"https://{keyVaultName}.vault.azure.net", UriKind.Absolute, out var keyVaultEndpointUri))
+        var vaultURI = configuration["OrchardCore:OrchardCore_KeyVault_Azure:VaultURI"];
+
+        Uri keyVaultEndpointUri;
+
+        if ((!string.IsNullOrWhiteSpace(vaultURI) && !Uri.TryCreate($"{vaultURI}", UriKind.Absolute, out keyVaultEndpointUri)) ||
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs:81]:
```suggestion
        if ((!string.IsNullOrWhiteSpace(vaultURI) && !Uri.TryCreate(vaultURI, UriKind.Absolute, out keyVaultEndpointUri)) || 
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs [hunk 2]
```diff
@@ -74,9 +74,15 @@ private static void AddOrchardCoreAzureKeyVault(
             throw new Exception("The 'KeyVaultName' property is no configured. Please configure it by specifying the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
         }
 
-        if (!Uri.TryCreate($"https://{keyVaultName}.vault.azure.net", UriKind.Absolute, out var keyVaultEndpointUri))
+        var vaultURI = configuration["OrchardCore:OrchardCore_KeyVault_Azure:VaultURI"];
+
+        Uri keyVaultEndpointUri;
+
+        if ((!string.IsNullOrWhiteSpace(vaultURI) && !Uri.TryCreate(vaultURI, UriKind.Absolute, out keyVaultEndpointUri)) ||
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs:81]: I think there is a logical error here and does not meet the comments in the app settings. I think if you provide the following settings, you'll get an exception:

```
"OrchardCore_KeyVault_Azure": {
    "KeyVaultName": "* << invalid name"
    "VaultURI": "https://abc.vault.azure.net"
}
```

I think you logic should look more like this:

```csharp
var vaultUri = configuration["OrchardCore:OrchardCore_KeyVault_Azure:VaultURI"];

Uri keyVaultEndpointUri;

if (!string.IsNullOrWhiteSpace(vaultUri) && !Uri.TryCreate(vaultUri, UriKind.Absolute, out keyVaultEndpointUri))
{
    throw new Exception("Invalid value used for 'VaultURI' property. Please provide a valid key-vault name using the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
} 
else 
{
    var keyVaultName = configuration["OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName"];

    if (string.IsNullOrEmpty(keyVaultName))
    {
        throw new Exception("The 'KeyVaultName' property is no configured. Please configure it by specifying the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
    }

    if (!Uri.TryCreate($"https://{keyVaultName}.vault.azure.net", UriKind.Absolute, out var keyVaultEndpointUri))
    {
        throw new Exception("Invalid value used for 'KeyVaultName' property. Please provide a valid key-vault name using the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
    }
}
```

I would even go an extra step an more that login into a separate private method using something like this:

```csharp
private Uri GetVaultHost()
{
  var vaultUri = configuration["OrchardCore:OrchardCore_KeyVault_Azure:VaultURI"];
  
  if (!string.IsNullOrWhiteSpace(vaultUri))
  {
      if(!Uri.TryCreate(vaultUri, UriKind.Absolute, out uri))
      {
          throw new Exception("Invalid value used for 'VaultURI' property. Please provide a valid key-vault name using the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
      }
  
      return uri;
  } 
  
    var keyVaultName = configuration["OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName"];
    
    if (string.IsNullOrEmpty(keyVaultName))
    {
        throw new Exception("The 'KeyVaultName' property is no configured. Please configure it by specifying the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
    }
    
    if (!Uri.TryCreate($"https://{keyVaultName}.vault.azure.net", UriKind.Absolute, out var host))
    {
        throw new Exception("Invalid value used for 'KeyVaultName' property. Please provide a valid key-vault name using the 'OrchardCore:OrchardCore_KeyVault_Azure:KeyVaultName' settings key.");
    }
    
    return host;
}

```

Comment (mohit-naroliya) [reply src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs:81]: @MikeAlhayek  - The approach using the GetVaultHost() method seems more appropriate. I'm making the necessary changes and will commit them again.

File: src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+using System;
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Configuration.KeyVault/Extensions/AzureKeyVaultConfigurationExtension.cs:1]: Why this was added?
```suggestion
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

**Review by MikeAlhayek — APPROVED** [review]

**Comment by sebastienros** [comment]
Thanks @mohit-naroliya. Could you also add the entry in the "mock" app settings file `src\OrchardCore.Cms.Web\appsettings.json` ?

**Pull request merged**